Refactoring, Consolidation, and Updates to the Web UI

pwendell opened the pull request:
This patch performs an internal re-factoring of the Spark Web UI and adds a new UI component that exposes stage progress and metrics.

The re-factoring removes the dependency on Spray and instead uses Scala's built in XML support and a lightweight server based on Jetty. The changes also consolidate code related to the spark ui in a "ui" package which can server as a point for future expansion. Certain very common rendering operations (such as enumerating a list of items) are factored out into utility functions.

The expansion adds a stage-level monitoring feature which exposes the progress of ongoing stages and statistical metrics for both ongoing and historical stages. It is meant to to be a modest first step towards a more comprehensive workflow UI. Nonetheless, it exposes some very useful previously hidden metrics, such as distribution data about tasks within a job, and an adjacent view of the service time and locality properties of given tasks.

So far this has been tested using a local standalone cluster with a variety of job conditions, including failures. I'm in the process of doing more rigorous tests, but wanted to put this up for review.

pwendell committed:
Partially working checkpoint

pwendell committed:
Completely hacked version of block manager UI in jetty

pwendell committed:
Minor cleanup

pwendell committed:
Adding deterministic port

pwendell committed:
Fix in storage status page

pwendell committed:
Spray refactoring for master web UI

pwendell committed:
Commenting unfinished part

pwendell committed:
Adding WebUI file

pwendell committed:
Cleaning up master web UI

pwendell committed:
Some comments

pwendell committed:
Adding worker version of UI

pwendell committed:
Changing JSON protocol and removing spray code

pwendell committed:
Some JSON cleanup

pwendell committed:
Removing dead code

pwendell committed:
Updating Maven build

pwendell committed:
Removing twirl files

pwendell committed:
Import cleanup

pwendell committed:
Refactoring UI packages

pwendell committed:
Stash

pwendell committed:
Initial steps towards job progress UI

pwendell committed:
Adding helper function to make listing tables

pwendell committed:
Refactoring into different modules

pwendell committed:
More work on percentile data:

pwendell committed:
Documentation and some refactoring

pwendell committed:
Some renaming and comments

pwendell committed:
Style cleanup

pwendell committed:
Showing shuffle status and purging old stages

pwendell committed:
Linking RDD information when available in stages

pwendell committed:
Using scala in XML imports

pwendell committed:
Some tweaking with shared page header

pwendell committed:
Handling entirely failed stages

pwendell committed:
Some cleanup of styling

pwendell committed:
Updating based on changes to JobLogger (and one small change to JobLogger)

pwendell committed:
Work around for initalization issue

pwendell committed:
Cleaning up some code for Job Progress

pwendell committed:
Minor style cleanup

pwendell committed:
A few bug fixes and a unit test

pwendell committed:
Allowing for requested port on construction

pwendell committed:
Adding another unit test to Web UI suite

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/154/

mateiz commented on core/pom.xml:
```diff
@@ -56,6 +56,10 @@
     </dependency>
     <dependency>
+      <groupId>net.liftweb</groupId>
+      <artifactId>lift-json_2.9.2</artifactId>
```
How self-contained is this, i.e., does it bring in a lot of other dependencies from Lift? You can use the sbt dependency-tree plugin to check this.

pwendell replied:
One thing that's nice about this is it has exactly one dependency: the paranamer utility. This only conflicts with one other library we use (avro), where it causes a slight version bump that should be fine.

mateiz commented on core/src/main/scala/spark/deploy/master/ui/ApplicationPage.scala:
```diff
@@ -0,0 +1,100 @@
+  def renderJson(request: HttpServletRequest): JValue = {
+    val appId = request.getParameter("appId")
+    val stateFuture = (master ? RequestMasterState)(timeout).mapTo[MasterState]
+    val state = Await.result(stateFuture, 3 seconds)
```
These 3-second timeouts are probably too low if there is GC; would be better to do 20-30s

mateiz commented on core/src/main/scala/spark/deploy/master/ui/MasterWebUI.scala:
```diff
@@ -0,0 +1,59 @@
+}
+
+object MasterWebUI {
+  val STATIC_RESOURCE_DIR = "spark/webui/static"
```
Wouldn't this be better as spark/ui/static, since we don't have a spark.webui package anymore?

mateiz commented on core/src/main/scala/spark/ui/SparkUI.scala:
```diff
@@ -0,0 +1,63 @@
+  private[spark] def appUIAddress = "http://" + host + ":" + boundPort.getOrElse("-1")
+}
+
+object SparkUI {
```
This and all the new objects/classes you made should probably be private[spark]; run sbt doc and look at the generated docs to see whether anything in this package shows up.

mateiz commented on core/src/main/scala/spark/ui/jobs/JobProgressUI.scala:
```diff
@@ -0,0 +1,124 @@
+
+private[spark] class JobProgressListener extends SparkListener {
+  // How many stages to remember
+  val RETAINED_STAGES = 1000
```
Probably should make this configurable

mateiz commented:
Thanks Patrick, this is really exciting! I've made a few quick comments on it but will continue looking.

pwendell committed:
Making all new classes package private

pwendell committed:
Moving static ui package

pwendell committed:
Bumping timeouts

pwendell committed:
Adding config option for retained stages

pwendell committed:
One more private class

pwendell commented:
Hey Matei,

Just added commits to address all your comments. Re: The lift-json library, it's very self-contained and is decoupled from the overall lift framework, that's why I chose it.

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/158/

mateiz commented:
Hey Patrick,

I tried it out a bit more and I think it looks great! A few observations though:
- If I click on an active stage, I get the following exception:
  
  13/06/27 10:22:00.668 WARN AbstractHttpConnection: /stages/stage/?id=6
  java.util.NoSuchElementException: key not found: 6
      at scala.collection.MapLike$class.default(MapLike.scala:225)
      at scala.collection.mutable.HashMap.default(HashMap.scala:45)
      at scala.collection.MapLike$class.apply(MapLike.scala:135)
      at scala.collection.mutable.HashMap.apply(HashMap.scala:45)
      at spark.ui.jobs.StagePage.render(StagePage.scala:22)
      at spark.ui.jobs.JobProgressUI$$anonfun$getHandlers$1.apply(JobProgressUI.scala:38)
      at spark.ui.jobs.JobProgressUI$$anonfun$getHandlers$1.apply(JobProgressUI.scala:38)
      at spark.ui.JettyUtils$$anon$1.handle(JettyUtils.scala:44)
- For some reason, the formatting of the body seems to be different from the previous UI; the font in tables is much larger than the body font size for Bootstrap should be, or the way it appears in other UI elements. (This is on Safari.)
- Tasks are showing a very low service time even when I make them sleep 1+ seconds each.
- Replace the name "service time" with "duration" to make it easier to understand
- For times displayed on the UI, it might be nice to have a user-understandable number such as "50 ms" or "52 s" instead of hh:mm:ss.sss. For JSON you should definitely standardize on one unit.
- It would be cool to render some progress bars where we show things like 0/10. Shouldn't be too hard with some nested divs or SVG.

mateiz commented:
Forgot to add one more cosmetic thing: the menu at the top with "Jobs" and "Storage" is kind of rough the way it is now.. would be better with some kind of tabs.

pwendell committed:
Fixing bug in local scheduler time recording

pwendell committed:
Handful of changes based on matei's review

- Avoid exception when no tasks have finished for a stage
- Adding DOCTYPE so css renders properly
- Adding progress slider

pwendell committed:
Adding better tabs for UI headers.

pwendell commented:
Hey Matei,

I've addressed everything you suggested in the review. The service time glitch was a bug in the local scheduler, that should be working now.

In order to add fancier headers, I had to create different codepaths for the Master/Worker UI generation and the Spark UI, this adds a bit more XML code in UIUtils. The new headers are based on the ones we use in the docs.

pwendell commented:
Jenkins, test this please.

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/166/

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/168/

pwendell committed:
Forcing Jetty to run as daemon

pwendell commented:
Jenkins, test this please.

mateiz commented:
Hey Patrick, the header still looks wrong on the SparkUI in Safari: there's an invalid image URL. Is that meant to be the logo?

mateiz commented on core/src/main/scala/spark/ui/jobs/IndexPage.scala:
```diff
@@ -0,0 +1,99 @@
+
+  def render(request: HttpServletRequest): Seq[Node] = {
+    val stageHeaders = Seq("Stage ID", "Origin", "Submitted", "Duration", "Progress",
+                           "Tasks: Complete/Total", "Shuffle Activity", "Stored RDD")
```
I'd merge the "progress" and "tasks: complete/total" into a single column here, just "progress". Then we can display a progress bar with the numbers next to it.

mateiz commented on core/src/main/scala/spark/ui/jobs/JobProgressUI.scala:
```diff
@@ -0,0 +1,125 @@
+private[spark] class JobProgressUI(val sc: SparkContext) {
+  private var _listener: Option[JobProgressListener] = None
+  def listener = _listener.get
+  val dateFmt = new SimpleDateFormat("EEE, MMM d yyyy HH:mm:ss")
```
Make this just "MMM d yyyy HH:mm:ss"

mateiz replied:
Or even "yyyy/MM/dd HH:mm:ss" to match the deploy UI one

mateiz commented on core/src/main/scala/spark/ui/jobs/IndexPage.scala:
```diff
@@ -0,0 +1,99 @@
+
+    val activeStageTable: NodeSeq = listingTable(stageHeaders, stageRow(), activeStages)
+    val completedStageTable = listingTable(stageHeaders, stageRow(), completedStages)
+    val failedStageTable: NodeSeq = listingTable(stageHeaders, stageRow(false), failedStages)
```
Why don't we want click-through links to failed stages? It seems those are specifically some stages where you'd like more info.

mateiz commented on core/src/main/scala/spark/ui/jobs/IndexPage.scala:
```diff
@@ -0,0 +1,99 @@
+    <tr>
+      {if (showLink) {<td><a href={"/stages/stage?id=%s".format(s.id)}>{s.id}</a></td>}
+       else {<td>{s.id}</td>}}
+      <td>{s.origin}</td>
```
IMO place the link around the origin, not the ID, so that it's easier to click.

pwendell replied:
Okay will do - I felt like the origin was a bit worse because it will be the same for multiple phases, so it's not good at distinguishing what you are clicking on. But I'll change it.

pwendell committed:
Missing logo file

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/169/

mateiz commented on core/src/main/scala/spark/ui/jobs/IndexPage.scala:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+
+  def makeSlider(completed: Int, total: Int): Seq[Node] = {
```
This should be called makeProgressBar actually since a slider usually means a draggable control in UI-speak

mateiz commented on core/src/main/scala/spark/ui/jobs/IndexPage.scala:
```diff
@@ -0,0 +1,99 @@
+      <rect width={width.toString} height={height.toString}
+            fill="white" stroke="black" stroke-width="1" />
+      <rect width={completeWidth.toString} height={height.toString}
+            fill="rgb(206,206,247)" stroke="black" stroke-width="1" />
```
Cosmetic thing: switch this to one of the theme colors.. maybe the blue for links; and consider making the border slightly lighter, maybe rgb(51, 51, 51) to match the text color. Ultimately it would be good to put these in a CSS file but I guess we can leave them here for now.

mateiz commented on core/src/main/scala/spark/ui/JettyUtils.scala:
```diff
@@ -0,0 +1,227 @@
+                    <ul id="infolist">
+                      <li>Master: <strong>{sc.master}</strong></li>
+                      <li>Application: <strong>{sc.appName}</strong></li>
+                      <li>Executors: <strong>{sc.getExecutorStorageStatus.size}</strong></li>
```
Put the app name before the master URL here

AmplabJenkins commented:
Thank you for your pull request. All automated tests for this request have passed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/172/

pwendell committed:
Minor changes from Matei's review

pwendell committed:
Styling for progress bar

pwendell committed:
Allowing details for failed stages

pwendell commented:
Hey Matei,

These commits should address everything in your feedback. I'd still like to manually test this a bit more, but I'm going away for the weekend - so wanted to post for review.
- Patrick

pwendell committed:
Print exception message

pwendell committed:
Adding test script

pwendell commented:
Hey Matei - this is "good to go" from my end. I added a testing script that you can use to generate interesting jobs for the UI.

mateiz commented on core/src/main/scala/spark/ui/JettyUtils.scala:
```diff
@@ -0,0 +1,227 @@
+      case Some(res) =>
+        staticHandler.setResourceBase(res.toString)
+      case None =>
+        logError("Could not find resource path for Web UI: " + resourceBase)
```
Shouldn't this throw an exception?

mateiz commented on core/src/main/scala/spark/ui/JettyUtils.scala:
```diff
@@ -0,0 +1,227 @@
+      }
+    }
+
+
```
Make this just one blank line

mateiz commented on core/src/main/scala/spark/ui/JettyUtils.scala:
```diff
@@ -0,0 +1,227 @@
+  }
+}
+
+object Page extends Enumeration { val Storage, Jobs = Value }
```
Make this private[spark] and move it to a separate file, Page.scala

mateiz commented on core/src/main/scala/spark/ui/JettyUtils.scala:
```diff
@@ -0,0 +1,227 @@
+object Page extends Enumeration { val Storage, Jobs = Value }
+
+/** Utility functions for generating XML pages with spark content. */
+private[spark] object UIUtils {
```
This should also be in a separate file

mateiz commented on core/src/main/scala/spark/ui/UIWorkloadGenerator.scala:
```diff
@@ -0,0 +1,61 @@
+ *
+ * Usage: ./run spark.ui.UIWorkloadGenerator [master]
+ */
+private[spark] object UIWorkloadGenerator {
```
This is great, very useful!

mateiz commented on core/src/main/scala/spark/deploy/worker/ui/WorkerWebUI.scala:
```diff
@@ -0,0 +1,74 @@
+    }
+  }
+
+  def log(request: HttpServletRequest): String = {
```
Viewing logs doesn't work right now -- I get

```
13/07/01 12:31:40.966 WARN AbstractHttpConnection: /log/?appId=app-20130701122932-0000&executorId=0&logType=stdout
java.lang.NullPointerException
    at spark.deploy.worker.ui.WorkerWebUI.log(WorkerWebUI.scala:59)
    at spark.deploy.worker.ui.WorkerWebUI$$anonfun$2.apply(WorkerWebUI.scala:37)
    at spark.deploy.worker.ui.WorkerWebUI$$anonfun$2.apply(WorkerWebUI.scala:37)
    at spark.ui.JettyUtils$$anon$1.handle(JettyUtils.scala:45)
    at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1040)
    at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:976)
    at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:135)
    at org.eclipse.jetty.server.handler.HandlerList.handle(HandlerList.java:52)
    at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:116)
    at org.eclipse.jetty.server.Server.handle(Server.java:363)
    at org.eclipse.jetty.server.AbstractHttpConnection.handleRequest(AbstractHttpConnection.java:483)
    at org.eclipse.jetty.server.AbstractHttpConnection.headerComplete(AbstractHttpConnection.java:920)
    at org.eclipse.jetty.server.AbstractHttpConnection$RequestHandler.headerComplete(AbstractHttpConnection.java:982)
    at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:635)
    at org.eclipse.jetty.http.HttpParser.parseAvailable(HttpParser.java:235)
    at org.eclipse.jetty.server.AsyncHttpConnection.handle(AsyncHttpConnection.java:82)
    at org.eclipse.jetty.io.nio.SelectChannelEndPoint.handle(SelectChannelEndPoint.java:628)
    at org.eclipse.jetty.io.nio.SelectChannelEndPoint$1.run(SelectChannelEndPoint.java:52)
    at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:608)
    at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:543)
    at java.lang.Thread.run(Thread.java:680)
```

mateiz commented on core/src/main/scala/spark/ui/jobs/IndexPage.scala:
```diff
@@ -0,0 +1,114 @@
+      <rect width={width.toString} height={height.toString}
+            fill="white" stroke="black" stroke-width="1" />
+      <rect width={completeWidth.toString} height={height.toString}
+            fill="rgb(51,51,51)" stroke="black" stroke-width="1" />
```
Small cosmetic thing.. make the progress bar the same color as the links on the page (whatever blue that is), and make the border rgb(51,51,51)

mateiz commented on core/src/main/scala/spark/Utils.scala:
```diff
@@ -486,6 +486,26 @@ private object Utils extends Logging {
+
+    ms match {
+      case t if t < second =>
+        "%dms".format(t)
```
Add a space before the units, i.e. "%d ms" or "%d s"

mateiz commented on core/src/main/scala/spark/Utils.scala:
```diff
@@ -486,6 +486,26 @@ private object Utils extends Logging {
+      case t if t < second =>
+        "%dms".format(t)
+      case t if t < minute =>
+        "%d.%03ds".format(t / second, t % second)
```
Here I'd put just one decimal place

mateiz commented on core/src/main/scala/spark/Utils.scala:
```diff
@@ -486,6 +486,26 @@ private object Utils extends Logging {
+      case t if t < minute =>
+        "%d.%03ds".format(t / second, t % second)
+      case t if t < hour =>
+        "%d:%02d".format(t / minute, (t % minute) / second)
```
It might be better to use "4.5 min" instead of "4:30", though I'm not sure what's standard; the reason is that the latter looks like a time of the day. Same for the hours -- "1.2 h" instead of "1:12:00" or whatever.

mateiz commented:
Hey Patrick, I took another look at this and made some comments. In particular, the log links on the standalone workers don't seem to be working.

One other weird thing I noticed is that failed tasks don't have a duration field. It would be good to put in the time they took until they failed, so the user knows whether it happened somewhere in the middle.

mateiz commented on core/src/main/scala/spark/deploy/worker/ui/WorkerWebUI.scala:
```diff
@@ -0,0 +1,74 @@
+    val path = "%s/%s/%s/%s".format(workDir.getPath, appId, executorId, logType)
+    val source = Source.fromFile(path)
+    val lines = source.mkString
+    source.close()
```
Also, reading the whole file is into a String is asking for trouble here. You should either use sendFile if Jetty has that functionality (basically let it stream the file to the client), or cut off the file and only return the last 4-8 KB.

JoshRosen replied:
If you wanted to generalize this, you could allow the user to specify a length and offset into the log as part of the request (which would fix [SPARK-608](https://spark-project.atlassian.net/browse/SPARK-608)).  This would support the `tail` use case and also enable fancy things like long-polling-based Javascript log viewers.  To more easily support `tail`, you could have the start offset work like a Python array index so you could pass -1024 to start reading from the last 1024 bytes of the log.

pwendell committed:
Various formatting changes

pwendell committed:
Fixing bug where logs aren't shown

pwendell committed:
Package refactoring

pwendell committed:
Throw exception for missing resource

pwendell commented:
Okay, these commits fix everything except the log issue. I'm going to look into that now, it just may be a bit tricky.

pwendell committed:
Adding truncation for log files

pwendell commented:
Okay now this adds truncation to the log files. I bumped it up to 100KB rather than 8, because the logs are pretty chatty and 8KB might not go back very far. This should address everything from the latest review.

pwendell committed:
Removing dead code

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/174/

mateiz commented:
Patrick, do you want to wait to include the duration for failed tasks too?

pwendell committed:
Enforcing correct sort order for formatted strings

pwendell committed:
Tracking some task metrics even during failures.

pwendell commented:
Hey Matei,

I've added collection of metrics for tasks which didn't complete. I also fixed a long-standing bug where the sort order in the UI is not correct for formatted outputs.
- Patrick

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/177/

mateiz commented:
Thanks Patrick! I've merged this manually since recent PRs introduced some small conflicts.

AmplabJenkins commented:
Thank you for submitting this pull request. Unfortunately, the automated tests for this request have failed.
Refer to this link for build results: http://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/178/
